building_sim_plugins/building_sim_common/include/building_sim_common/door_common.hpp

```diff
@@ -0,0 +1,244 @@
+  // get_sdf_param functions in utils.hpp
+  auto sdf_clone = sdf->Clone();
+
+  MotionParams params;
+  get_sdf_param_if_available<double>(sdf_clone, "v_max_door", params.v_max);
+  get_sdf_param_if_available<double>(sdf_clone, "a_max_door", params.a_max);
+  get_sdf_param_if_available<double>(sdf_clone, "a_nom_door", params.a_nom);
+  get_sdf_param_if_available<double>(sdf_clone, "dx_min_door", params.dx_min);
+  get_sdf_param_if_available<double>(sdf_clone, "f_max_door", params.f_max);
+
+  auto door_element = sdf_clone;
+  std::string left_door_joint_name;
+  std::string right_door_joint_name;
+  std::string door_type;
+
+    // Get the joint names and door type
+  if (!get_element_required(sdf_clone, "door", door_element) ||
+    !get_sdf_attribute_required<std::string>(
+      door_element, "left_joint_name", left_door_joint_name) ||
+    !get_sdf_attribute_required<std::string>(
+      door_element, "right_joint_name", right_door_joint_name) ||
+    !get_sdf_attribute_required<std::string>(
+      door_element, "type", door_type))
+  {
+    RCLCPP_ERROR(node->get_logger(),
+      " -- Missing required parameters for [%s] plugin",
+      door_name.c_str());
+    return nullptr;
+  }
+
+  if ((left_door_joint_name == "empty_joint" &&
+      right_door_joint_name == "empty_joint") ||
+      (left_door_joint_name.empty() && right_door_joint_name.empty()))
+  {
+    RCLCPP_ERROR(node->get_logger(),
+      " -- Both door joint names are missing for [%s] plugin, at least one"
+      " is required", door_name.c_str());
+    return nullptr;
+  }
+
+    // Default values for joint limts
+  double left_joint_lower_limit = -1.57;
+  double left_joint_upper_limit = 0.0;
+  double right_joint_lower_limit = 0.0;
+  double right_joint_upper_limit = 1.57;
+
+  auto extract_limits = [&](SdfPtrT& joint_sdf)
+  {
+    auto joint_sdf_clone = joint_sdf->Clone();
+    std::string joint_name;
+    get_sdf_attribute_required<std::string>(
+      joint_sdf_clone, "name", joint_name);
+    auto element = joint_sdf_clone;
+    get_element_required(joint_sdf_clone, "axis", element);
+    get_element_required(element, "limit", element);
+    if (joint_name == left_door_joint_name)
+    {
+      get_sdf_param_if_available<double>(element, "lower", left_joint_lower_limit);
+      get_sdf_param_if_available<double>(element, "upper", left_joint_upper_limit);
+      RCLCPP_INFO(node->get_logger(),
+        "Joint [%s] lower [%f] upper [%f]",
+        joint_name.c_str(),left_joint_lower_limit, left_joint_upper_limit);
+    }
+    else if (joint_name == right_door_joint_name)
+    {
+      get_sdf_param_if_available<double>(element, "lower", right_joint_lower_limit);
+      get_sdf_param_if_available<double>(element, "upper", right_joint_upper_limit);
+      RCLCPP_INFO(node->get_logger(),
+        "Joint [%s] lower [%f] upper [%f]",
+        joint_name.c_str(),left_joint_lower_limit, left_joint_upper_limit);
+    }
+  };
```

Comment:
Would there be a way to remove the duplication here and make the lambda a function that returns the limits instead of modifying global variables?
I would personally remove the print statement as well but no strong feelings there.

Reply:
I've removed the duplication. The `make()` function is responsible for passing an `std::unordered_map<std::string, DoorCommon::DoorElement>` object to the constructor. The lambda now helps to directly populate this object for valid joint names. I think having the lambda modify the global object is more efficient than returning a pair and inserting it later. It also handles cases where the `joint name` attribute is not found in the `unordered_set<std::string>`. Else we would need to rely on pointers for the return value.